diff --git a/madifx.c b/madifx.c
index 2b90e5c..312aec6 100644
--- a/madifx.c
+++ b/madifx.c
@@ -3995,11 +3995,22 @@ static void __devexit snd_madifx_remove(struct pci_dev *pci)
 	pci_set_drvdata(pci, NULL);
 }
 
-static struct pci_driver madifx_driver = {
+static struct pci_driver driver = {
 	.name = KBUILD_MODNAME,
 	.id_table = snd_madifx_ids,
 	.probe = snd_madifx_probe,
 	.remove = __devexit_p(snd_madifx_remove),
 };
 
-module_pci_driver(madifx_driver);
+static int __init alsa_card_hdspm_init(void)
+{
+   return pci_register_driver(&driver);
+}
+
+static void __exit alsa_card_hdspm_exit(void)
+{
+   pci_unregister_driver(&driver);
+}
+
+module_init(alsa_card_hdspm_init)
+module_exit(alsa_card_hdspm_exit)
